### PR TITLE
Add Tag type

### LINKED ISSUE #178: Improve tag string initialization
I'd like a better way to initialize and store tags used in file processing, such as:
```cpp
const std::array<char, 4> TagVOL_{ 'V', 'O', 'L', ' ' }; // Volume file tag
```
In particular, it would be nice if we could use a string literal `"VOL "`, while also maintaining a strictly controlled struct size of 4 bytes.

----

I went back to the StackOverflow question about [initializing a `std::array<char, N>` with a string literal, and omitting the trailing null](https://stackoverflow.com/questions/33484233/how-to-initialize-a-stdarraychar-n-with-a-string-literal-omitting-the-trail), and did some reading and experimentation. What was posted works, and by adding a few `constexpr`, I was able to get g++ to do compile time initialization of `static const` variables. I think this would be really good for what we are doing here.

I'm about to take a break, so for now I'll post my experiment:
```cpp
#include <array>
#include <utility>
#include <iostream>


template<std::size_t N, std::size_t... IndexSequence>
constexpr std::array<char, N - 1> stripNullTerminator(const char (&string)[N], std::index_sequence<IndexSequence...>) {
  return {{string[IndexSequence]...}};
}

template <std::size_t N>
constexpr std::array<char, N - 1> stripNullTerminator(const char (&string)[N]) {
  return stripNullTerminator(string, std::make_index_sequence<N - 1>());
}


template<std::size_t N>
class NonNullTerminatedString {
public:
  constexpr NonNullTerminatedString(const char (&string)[N + 1]) : string(stripNullTerminator(string)) {}
  operator std::string() const {
    return std::string(string.data(), string.size());
  }
  friend std::ostream& operator << (std::ostream& out, const NonNullTerminatedString& nonNullTerminatedString) {
    return out.write(nonNullTerminatedString.string.data(), nonNullTerminatedString.string.size());
  }
private:
  std::array<char, N> string;
};


using Tag = NonNullTerminatedString<4>;


constexpr auto TagVol_ = Tag("VOL ");
constexpr auto TagVolh = Tag("volh");
constexpr auto TagVols = Tag("vols");
constexpr auto TagVblk = Tag("VBLK");
constexpr std::array<char, 4> tag2 = {'V', 'O', 'L', ' '};
const std::string s2(tag2.data(), tag2.size());


int main() {
  auto a = stripNullTerminator("Some string");
  std::cout << a.size() << std::endl;
  std::cout << s2 << std::endl << std::endl;

  std::cout << TagVol_ << std::endl;
  std::cout << TagVolh << std::endl;
  std::cout << TagVols << std::endl;
  std::cout << TagVblk << std::endl;

  return 0;
}
```

I'm uncertain if I really want to wrap `std::array` in a new type. I'm uncertain if I want the template helper methods global or wrapped in a new type or namespace. I'm uncertain if the recursive templates and parameter packs are really necessary, or if there might be a more clear way to implement this. I believe the variables could be declared as `const` rather than `constexpr`, though using `constexpr` may provide additional checks, and may change output.

I may also want to re-read and perhaps summarize some of the relevant articles. In particular:
[Parameter pack](https://en.cppreference.com/w/cpp/language/parameter_pack)
[`std::integer_sequence` and `std::index_sequence`](https://en.cppreference.com/w/cpp/utility/integer_sequence)
[`constexpr`](https://en.cppreference.com/w/cpp/language/constexpr)
[LiteralType](https://en.cppreference.com/w/cpp/named_req/LiteralType)
[RIP index_sequence, 2014-2017](https://medium.com/@matt.aubury/rip-index-sequence-2014-2017-9cc854aaad0)

### PR DESCRIPTION
**Pull request by DanRStevens**
This ~~relates to Issue~~ closes #178.

~~I'm still trying out a few ideas, but I figured I should get this up now for review and for CI purposes.~~ Done.

### CODE REVIEW

**Comment by DanRStevens** [comment]
Ok, I think I'm done playing around with this. The final commit is a bit of an experiment to compare having a full `Tag` class, as opposed to a simple `using` `Tag` alias. It adds a bit more code to do it that way, and the result is largely similar, though it does perhaps lead to a cleaner global namespace, and allow some additional conveniences, such as automatic conversion to `std::string`.

**Comment by Brett208** [comment]
I added the new source code files to the VS project files so the tag test cases are running now on MSVC builds. I've tested by extracting from and creating both CLM and VOL files. No problems noticed yet.

I still need to test creating/reading bitmap files since this code affects bitmaps.

I ran into another bug in XFile that I will push before continuing testing. Bug is unrelated to the TAG code, just found it when testing.

-Brett

File: OP2Utility.vcxproj.filters [hunk 1]
```diff
@@ -177,9 +177,12 @@
     <ClInclude Include="src\Map\SavedGameUnits.h">
       <Filter>Map</Filter>
     </ClInclude>
+    <ClInclude Include="src\Stream\SliceReader.h">
+      <Filter>Stream</Filter>
+    </ClInclude>
     <ClInclude Include="src\Tag.h">
       <Filter>Header Files</Filter>
-    </ClInclude>
+	</ClInclude>
```
Comment (DanRStevens) [inline OP2Utility.vcxproj.filters]: Hmm, looks like there is a tab/space inconsistency here. I wonder how that got introduced.

Looking into this, I found another inconsistency in the Test project file filters as well.

I'll update.

File: src/Tag.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+	// Allow construction from string literals
+	constexpr Tag(const char (&text)[5]) :
+		text(RemoveLastElement(text))
+	{}
+
+	// Allow construction from other Tag objects
+	constexpr Tag(const Tag& other) = default;
+
+	// Assignable
+	// Tag& operator =(const Tag& rhs) = default;
```
[the remark this answers is not in the document]
Comment (DanRStevens) [reply src/Tag.h]: Good catch. That should probably be removed. The case that was added for may be covered by the copy constructor. I had commented it out while testing to see if it was actually needed, and then forgotten about it.

**Comment by Brett208** [comment]
When I fixed the merge conflicts I used notepad++ instead of VS, so probably my fault on the spacing.

**Pull request merged**